lua/markdown.lua

```diff
@@ -8,22 +8,79 @@ local regex = {
     ordered_list        = "^%s*%d+[%)%.]",
 }
 
+local function get_nearest_ancestor_node(node, type)
+    if not node then
+        return nil
+    end
+    local cur_node = node
+    repeat
+        cur_node = cur_node:parent()
+    until cur_node == nil or cur_node:type() == type
+    return cur_node
+end
+
+function M.reenumerate_numbered_list()
+    local list_node = get_nearest_ancestor_node(vim.treesitter.get_node(), 'list')
+    if list_node == nil then
+        return
+    end
+    repeat
+        local idx = 1
+        for item, _ in list_node:iter_children()
+        do
+            local node = item:named_child { 0 }
+            if node:type():sub(1, 11) == 'list_marker' then
+                local sr, sc, er, ec = node:range()
+                local new_list_marker, _ = vim.api.nvim_buf_get_text(0, sr, sc, er, ec, {})[1]:gsub('%d+', idx)
+                vim.api.nvim_buf_set_text(0, sr, sc, er, ec, { new_list_marker })
+                idx = idx + 1
+            end
+        end
+        list_node = get_nearest_ancestor_node(list_node, 'list')
+    until not list_node
+end
+
+function M.deindent_list_item()
+    local cur_node = vim.treesitter.get_node()
+    local list_node = get_nearest_ancestor_node(cur_node, 'list')
+    local parent_list_node
+    if list_node then
+        parent_list_node = get_nearest_ancestor_node(list_node, 'list')
+        if not parent_list_node then
+            return
+        end
+    else
+        return
+    end
+    local psr, psc, per, pec = parent_list_node:named_child({ 0 }):named_child({ 0 }):range()
+    local parent_marker = vim.api.nvim_buf_get_text(0, psr, psc, per, pec, {})[1]
+    local csr, csc, cer, cec
+    if cur_node:type() ~= 'list_item' then
+        csr, csc, cer, cec = get_nearest_ancestor_node(cur_node, 'list_item'):named_child({ 0 }):range()
+    else
+        csr, csc, cer, cec = cur_node:named_child({ 0 }):range()
+    end
+    assert(psc <= csc)
+    vim.api.nvim_buf_set_text(0, csr, psc, cer, cec, { parent_marker })
+    print(parent_marker)
+    M.reenum_wrapper()
+end
+
 -- to make M.backspace only trigger after a new line has been created
-local should_run_callback = false
+local markdown_ns_id = vim.api.nvim_create_namespace('markdown')
 local function key_callback(key)
     local backspace_term = vim.api.nvim_replace_termcodes("<BS>",true, true, true)
 
     -- It sends some key on o and O "<80><fd>h", which is some special key I didn't ask for.
-    if should_run_callback and not (key:len() == 3 and key ~= backspace_term) then
+    if not (key:len() == 3 and key ~= backspace_term) then
         if key == backspace_term then
             M.backspace()
+            M.reenum_wrapper()
         end
-        should_run_callback = false
     end
+    vim.on_key(nil, markdown_ns_id)
```

Comment:
This on_key should be moved up one line into the first if, then the vim.schedule can be removed from the on_key call further down. The function of the schedule is to execute the callback after the key mentiond in the comment above has passed.